### ISSUE TITLE
Running Encoder and Decoder model separately

### ISSUE THREAD

**Issue by rojinbakhti** [ggml-org/whisper.cpp#1801]
Hi,
Can I use the API's to convert and run my whisper encoder and decoder models separately? If yes, how do I do this?

**Comment by bobqianic** [comment]
Yes, that's possible. However, you can't run the encoder with a small model and then switch to using a large model for the decoder. This is because the dimensions of the cross-attention tensors won't be compatible between the two different model sizes.

https://github.com/ggerganov/whisper.cpp/blob/1cf679dec4eca99aeaed4fe09a8092803bdecfc1/whisper.h#L259-L293

**Comment by rojinbakhti** [comment]
Thanks for the repluy, If I have my encoder output (hidden states) values in a float array, let's say from running the encoder part using tensorflow api's, is there a way to feed it into the whisper_decode? (ie calling whisper_decode wihout calling whisper_encode first, assuming I have the encoder output + tokens I want to feed the decoder beforehand).

**Comment by rojinbakhti** [comment]
@bobqianic Hi, have you had a chance to look at this question?
Thanks.

**Comment by rojinbakhti** [comment]
@ggerganov

**Comment by bobqianic** [comment]
> @bobqianic Hi, have you had a chance to look at this question? Thanks.

You'll need to manually modify `wstate.kv_cross`. Essentially, this portion of the encoder's code is designed to directly copy the intermediate tensors `Kcross` and `Vcross` into `wstate.kv_cross.k` and `wstate.kv_cross.v`. Otherwise, the intermediate tensor would be overwritten during subsequent executions.

https://github.com/ggerganov/whisper.cpp/blob/7a74e929c842489010f641156f2a5ac733b17016/whisper.cpp#L2099-L2108

In this section of the decoder's code, `wstate.kv_cross.k` and `wstate.kv_cross.v` are used.

https://github.com/ggerganov/whisper.cpp/blob/7a74e929c842489010f641156f2a5ac733b17016/whisper.cpp#L2411-L2434

**Comment by bobqianic** [comment]
`whisper_init_from_file_with_params()` -> `whisper_init_state()` -> `modify state` -> `whisper_decode_with_state()`

**Comment by ggerganov** [comment]
@rojinbakhti As @bobqianic correctly explained, it's not possible because the cross-attention KV cache will not be populated if you feed the encoder output directly to the `whisper_decode()`. One option to make this possible is to expose a `whisper_cross()` API that precomputes cross-attention cache. It shouldn't be a difficult change, but I'm not sure if I'll be able to take a look soon. Hopefully someone from the community can give it a try

> Otherwise, the intermediate tensor would be overwritten during subsequent executions.

@bobqianic Not only that, but this allows to call `whisper_decode()` multiple times and reuse the cross attention instead of computing it each time
